### ISSUE TITLE
Internal CLR error (0x80131506) with covariant return types

### ISSUE THREAD

**Issue by Dotnet-And-Happiness** [dotnet/runtime#46519]
### Description
```
using System;

namespace Test
{
    class Program
    {
        static void Main(string[] args)
        {
            var b =new AChild<BChild>(new BChild()).ReturnB();
	}
    }
	
  public class B{}

  public class BChild : B {}

  public abstract class A 
  {
    public abstract B ReturnB();
  }
  
  
  public class AChild<T> : A
  where T:B
  {
    public T Prop {get;set;}
    public AChild(T b)
    {
	    Prop = b;
    }
    public override T ReturnB()
    {
	    return Prop;
    }
  }
}
```
this code cause Internal CLR error (0x80131506) on Windows 10 x64 and segmentation fault on  linux(Fedora 33 x64)

if I change AChild class to:

```
public class AChild<T> : A
where T:B
{
	public T Prop {get;set;}
	public AChild(T b)
	{
		Prop = b;
	}
	public override B ReturnB()
	{
		return Prop;
	}
}
```
everything will work fine

.Net version - 5.0.101

**Comment by jkotas** [comment]
cc @janvorli

**Comment by janvorli** [comment]
This works fine in master, so it should be fixed in the upcoming 5.0.2 release. Dup of #45082.
